h2o-core/src/main/java/water/UDPRebooted.java

```diff
@@ -24,15 +25,45 @@
 
     public void send(H2ONode target) {
       assert this != none;
-      new AutoBuffer(target,udp.rebooted._prior).putUdp(udp.rebooted).put1(ordinal()).close();
+      // Note! To ensure that H2O version without the PUBDEV-4959 fix does not bring H2O with this fix into some unwanted
+      // state we need to first discover if we are indeed receiving shutdown packet from a H2O version with this fix.
+      // For this, we overload this first byte which is sent in both versions and contain ordinal number of the request type.
+      // If we choose number different than the possible ordinal number we can safely discover on which version we are running.
+
+      // When we discover that we run on a new version we can check if
+      // the shutdown request comes from the node in the current cluster
+      // otherwise we just ignore the request
+      new AutoBuffer(target,udp.rebooted._prior)
+              .putUdp(udp.rebooted)
+              .put1(MAGIC_SAFE_CLUSTER_KILL_BYTE)
+              .put1(ordinal())
+              .putInt(H2O.SELF._heartbeat._cloud_name_hash)
+              .close();
     }
     void broadcast() { send(H2O.SELF); }
   }
-
   static void checkForSuicide(int first_byte, AutoBuffer ab) {
     if( first_byte != UDP.udp.rebooted.ordinal() ) return;
-    int type = ab.get1();
-    suicide( T.values()[type], ab._h2o);
+    int shutdownPacketType = ab.get1();
+    if(shutdownPacketType == MAGIC_SAFE_CLUSTER_KILL_BYTE) { // we are running on a version with PUBDEV-4959 fix
+      shutdownPacketType = ab.get1(); // read the real type
+      int cloud_name_hash_origin = ab.getInt();
+      if (cloud_name_hash_origin == H2O.SELF._heartbeat._cloud_name_hash) {
+        suicide(T.values()[shutdownPacketType], ab._h2o);
+      }else {
+        for (H2OListenerExtension ext : ExtensionManager.getInstance().getListenerExtensions()) {
```

Comment:
this loop should be somewhere else, better API would just let you do something like

ListenerService.report(...)

and hide the implementation details in this service

Reply:
perfect! done